File: src/main/java/com/fengsheng/handler/remove_one_position_tos.java
package com.fengsheng.handler;

import com.fengsheng.Game;
import com.fengsheng.GameExecutor;
import com.fengsheng.HumanPlayer;
import com.fengsheng.Player;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

public class remove_one_position_tos extends AbstractProtoHandler<Fengsheng.remove_one_position_tos> {
    private static final Logger log = Logger.getLogger(remove_one_position_tos.class);

    @Override
    protected void handle0(HumanPlayer player, Fengsheng.remove_one_position_tos pb) {
        synchronized (Game.class) {
            if (player.getGame().isStarted()) {
                log.error("game already started");
                return;
            }
            Player[] oldPlayers = player.getGame().getPlayers();
            if (oldPlayers.length <= 2) return;
            int i = oldPlayers.length - 1;
            for (; i >= 0; i--)
                if (oldPlayers[i] == null) break;
            int nullIndex = i;
            Player[] players = new Player[oldPlayers.length - 1];
            System.arraycopy(oldPlayers, 0, players, 0, i);
            System.arraycopy(oldPlayers, i + 1, players, i, oldPlayers.length - i - 1);
            for (i = 0; i < players.length; i++) {
                if (players[i] != null)
                    players[i].setLocation(i);
            }
            player.getGame().setPlayers(players);
            for (Player p : players) {
                if (p instanceof HumanPlayer)
                    ((HumanPlayer) p).send(Fengsheng.remove_one_position_toc.newBuilder().setPosition(nullIndex).build());
            }
            for (Player p : players)
                if (p == null) return;
            log.info("已满" + players.length + "个人，游戏开始。。。");
            player.getGame().setStarted(true);
            GameExecutor.post(player.getGame(), player.getGame()::start);
            Game.newInstance();
        }
    }
}

File: src/main/kotlin/handler/remove_one_position_tos.kt
package com.fengsheng.handler

import com.fengsheng.Config
import com.fengsheng.HumanPlayer
import com.fengsheng.protos.Fengsheng
import org.apache.log4j.Logger

class remove_one_position_tos : AbstractProtoHandler<Fengsheng.remove_one_position_tos>() {
    override fun handle0(r: HumanPlayer, pb: Fengsheng.remove_one_position_tos) {
        if (r.game!!.isStarted) {
            log.error("game already started")
            r.sendErrorMessage("游戏已经开始了")
            return
        }
        val oldPlayers = r.game!!.players
        if (oldPlayers.size <= 2) {
            r.sendErrorMessage("至少2人局")
            return
        }
        if (!Config.IsGmEnable) {
            val humanCount = oldPlayers.count { it is HumanPlayer }
            if (humanCount >= 3 && oldPlayers.size <= 5) {
                r.sendErrorMessage("至少5人局")
                return
            }
        }
        val index = oldPlayers.indexOfFirst { p -> p == null }
        val players = oldPlayers.filterIndexed { i, _ -> i != index }.toTypedArray()
        r.game!!.players = players
        players.forEachIndexed { i, p ->
            p?.location = i
            if (p is HumanPlayer) p.send(Fengsheng.remove_one_position_toc.newBuilder().setPosition(index).build())
        }
        for (p in players)
            if (p == null) return
        log.info("已满${players.size}个人，游戏开始。。。")
        r.game!!.setStartTimer()
    }

    companion object {
        private val log = Logger.getLogger(remove_one_position_tos::class.java)
    }
}